# Project Outline Notebook


## 1. Customer Behaviour Profiling & Micro-Segmentation

## 2. Problem Statement & Business Goal

### Problem Statement

Retail businesses generate large amounts of transactional customer data, but identifying meaningful customer groups and understanding shopping behavior remains a challenge. This project aims to analyze customer purchasing patterns and perform micro-segmentation using clustering techniques based on RFM metrics, Average Basket Size, and Product Diversity to identify distinct customer behavior profiles.

### Business Goals

- Identify customer segments based on purchasing behavior  
- Understand customer spending and engagement patterns  
- Profile high-value, loyal, and churn-risk customers  
- Support targeted marketing and personalized offers  
- Improve customer retention and business decision-making  

## 3. Dataset Details

- **Dataset Name:** UCI Online Retail II Data Set  
- **Source:** Kaggle
- **Dataset Link:** https://www.kaggle.com/datasets/jillwang87/online-retail-ii  

In [1]:
!pip install kagglehub

In [2]:
# Import KaggleHub
import kagglehub

In [3]:
import pandas as pd
path = kagglehub.dataset_download("jillwang87/online-retail-ii")
df = pd.read_csv(path+"/online_retail_09_10.csv")

100%|██████████| 14.2M/14.2M [00:01<00:00, 10.7MB/s]

Extracting files...


## 4. Initial Dataset Inspection

In [6]:
print(df.columns)
print(df.dtypes)
print(df.notnull().sum())
print(df.isnull().sum())
print(df.nunique())

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')
InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object
InvoiceNo      525461
StockCode      525461
Description    522533
Quantity       525461
InvoiceDate    525461
UnitPrice      525461
CustomerID     417534
Country        525461
dtype: int64
InvoiceNo           0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     107927
Country             0
dtype: int64
InvoiceNo      28816
StockCode       4632
Description     4681
Quantity         825
InvoiceDate    25296
UnitPrice       1606
CustomerID      4383
Country           40
dtype: int64


## 5. Initial Understanding & Observations

- The dataset contains **525,461 retail transaction records** with 8 features related to invoices, products, customers, pricing, and geographic information.

- The dataset includes both **categorical** and **numerical** variables:
  - Numerical: Quantity, UnitPrice
  - Categorical: InvoiceNo, StockCode, Description, CustomerID, Country

- `InvoiceDate` is currently stored as an object datatype and requires conversion to datetime format for time-based analysis such as Recency calculation and purchase trend analysis.

- `CustomerID` contains **107,927 missing values**, which is significant because customer-level behavioral profiling and segmentation depend heavily on customer identification.

- `Description` contains **2,928 missing values**, indicating some products lack descriptions, which may affect product-level analysis.

- Other columns have no missing values, indicating relatively stable transactional information.

- The dataset contains:
  - 28,816 unique invoices
  - 4,632 unique products
  - 4,383 unique customers
  - Customers from 40 different countries

- The high number of unique invoices and products indicates rich transactional diversity suitable for customer behavior analysis and micro-segmentation.

- Multiple purchases by the same customers suggest the dataset is appropriate for RFM analysis and customer profiling.

- The dataset appears to represent real-world retail transactional data, which may include cancellations, returns, duplicate transactions, and noisy records that require preprocessing before clustering and behavioral analysis.

- The presence of transactional, customer, and product-level information makes the dataset highly suitable for:
  - Customer segmentation
  - Behavioral profiling
  - RFM analysis
  - Spending pattern analysis
  - Customer retention analysis
  - Personalized marketing insights

## 6. Proposed Workflow

Data Loading → Data Understanding & Preprocessing → Exploratory Data Analysis (EDA) → Feature Engineering (RFM, Average Basket Size, Product Diversity) → Customer Behaviour Profiling → Customer Micro-Segmentation using Clustering → Cluster Evaluation & Interpretation → Business Insights & Visualization

## 7. Tools & Technologies

- Python
- Pandas
- NumPy
- Matplotlib
- Seaborn
- Scikit-learn
- Jupyter Notebook

## 8. References

1. Kaggle – UCI Online Retail II Dataset  
   https://www.kaggle.com/datasets/jillwang87/online-retail-ii
